src/storage/src/upsert_continual_feedback.rs

```diff
@@ -0,0 +1,826 @@
+
+                        match persist_event {
+                            AsyncEvent::Data(_cap, data) => {
+                                persist_stash.extend(data.into_iter().map(|((key, value), ts, diff)| {
+                                    (key, value, ts, diff)
+                                }))
+                            }
+                            AsyncEvent::Progress(upper) => persist_upper = upper,
+                        }
+                    }
+
+                    if hydrating {
+
+                        // Determine if this is the last time we call
+                        // consolidate_snapshot_chunk, and update our
+                        // `hydrating` state.
+                        //
+                        // There are two situations in which we're reading a
+                        // snapshot:
+                        //
+                        // 1. Reading the initial snapshot from the source: in
+                        //    this case there is no state in our output yet that
+                        //    we could read back in and our resume_upper is [0].
+                        //    We know we're done once the persist_upper is
+                        //    "past" that.
+                        // 2. Re-hydrating our state from persist: some instance
+                        //    of the source has already read/ingested/written
+                        //    down the initial snapshot from the source. We are
+                        //    reading a snapshot of our own upsert state back in
+                        //    from persist. In this case we read the snapshot at
+                        //    a timestamp that is "before" the resume upper, and
+                        //    we know that we're done when the persist_upper is
+                        //    "at or past" the resume_upper.
+                        //
+                        // It's unfortunate that we have to differentiate
+                        // between these cases, both here and in code below. The
+                        // root of the problem is that when reading the initial
+                        // source snapshot, the `resume_upper` is `[0]`, but
+                        // we're also reading in those updates at time `0`. When
+                        // starting the operator at any other `resume_upper`,
+                        // updates are read at a timestamp before that.
+                        let last_snapshot_chunk =  if resume_upper.less_equal(&G::Timestamp::minimum()) {
+                            PartialOrder::less_than(&resume_upper, &persist_upper)
+                        } else {
+                            PartialOrder::less_equal(&resume_upper, &persist_upper)
+                        };
+
+                        // Sort by ts and only present updates that are not
+                        // beyond the resume upper.
+                        persist_stash.sort_unstable_by(|a, b| {
```

Comment:
This will do at least one pass of the data to sort it (`n logn` in the worst case) and then below you do a `drain(..idx)` which will perform a big memmov of everything that comes after `idx` to the beginning of the vector. You'd be better off just using `drain_filter_swapping` on the data without worrying about sorting them (the time is ignored anyway).

Reply:
This code comes straight from https://github.com/MaterializeInc/materialize/blob/2280405a2e1f8a44fa1c8f046a718c013ff7af6b/src/storage/src/upsert.rs#L487. I can look into changing this as well, but at some point I might have to do them as a follow-up change.